### PR TITLE
Accept HTTP 202 as a valid response when posting metrics

### PR DESCRIPTION
**Pull request by jamesmulcahy**

### CODE REVIEW

**Comment by jamesmulcahy** [comment]
@dmuino I'm seeing 202 response being logged as errors:

```
{"level":"error","time":"2021-05-14T18:23:39Z","message":"Could not POST measurements: HTTP 202 <nil>"}
```

Granted, this error is from v0.1.1, but the logic is the same in the latest version -- they both assume that the client will only receive 200s.

File: registry.go [hunk 1]
```diff
@@ -264,7 +264,7 @@ func (r *Registry) sendBatch(measurements []Measurement) {
 		} else {
 			sent := int64(0)
 			dropped := int64(0)
-			if resp.status == 200 {
+			if resp.status == 200 || resp.status == 202 {
```
Comment (dmuino) [inline registry.go:267]: Makes sense but I'd update it so all 2xx are considered valid. Something like `if resp.status / 100 == 2`.

Note the test `TestRegistry_publish_errors` (file `registry_test.go` needs to be updated as well).

@copperlight is now the maintainer of the spectator thin clients, so adding him to the review.

Comment (brharrington) [reply registry.go:267]: Note, a 200 and 202 should be the only 2xx responses that are used. A 202 indicates a partial failure and that some datapoints in the payload were invalid and rejected.

**Comment by brharrington** [comment]
Looks like spectator-go was never updated to handle validation responses, e.g. https://github.com/Netflix/spectator-js/pull/39. I suppose it could be updated to support that, but not sure that is worth the effort vs working on moving this library to use spectatord.

**Review by copperlight — COMMENTED** [review]
As Brian noted, 202 is a partial failure response, and should not be treated the same as a 200 response. The existing implementation of status code handling in this client is consistent with the implementation in spectator-js, which we consider to be a reference for the other non-Java clients:

https://github.com/Netflix/spectator-go/blob/v0.1.2/registry.go#L267-L293

https://github.com/Netflix/spectator-js/blob/master/src/publisher.js#L142-L174

All non-200 responses (that are not server errors) attempt to parse an error response and increment the number of metrics sent with the number of metrics in the batch, minus the number of metrics that were dropped.

There is difference between `v0.1.1` and `v0.1.2` with respect to handling these response codes - the latest version has the correct implementation, which matches spectator-js. You should upgrade to the latest.

At some point, we will want to move this client to spectatord, but there are a number of other competing priorities which preclude starting work on that.

Given that the existing implementation is correct, we will decline this PR.

**Comment by jamesmulcahy** [comment]
Thanks @copperlight, @brharrington -- great context.  I hadn't appreciated that 202 meant things were being dropped.

The new client version is much more helpful in this regard too; the error output points me to the exact problem, which I'll go and fix now.  Thanks!

**Pull request closed without merging**